File: native/ios/Sources/RegistrationViews.swift
import SwiftUI

struct PolicyLinks: View {
    let policies: MobilePolicies
    let content: MobileContent
    @State private var selected: String?
    var body: some View {
        VStack(alignment: .leading) {
            ForEach(policies.pages.keys.sorted(), id: \.self) { key in
                Button(policies.pages[key]?.title ?? key) { selected = key }
                    .accessibilityIdentifier("policy-\(key)")
            }
        }
        .sheet(isPresented: Binding(get: { selected != nil }, set: { if !$0 { selected = nil } })) {
            if let key = selected, let page = policies.pages[key] {
                PolicyReader(page: page, policies: policies, content: content) { selected = nil }.id(key)
            }
        }
    }
}

private struct PolicyReader: View {
    let page: PolicyPage
    let policies: MobilePolicies
    let content: MobileContent
    let close: () -> Void
    @Environment(\.openURL) private var openURL
    @State private var linkFailed = false
    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 20) {
                    Text("\(content.text("policyUpdated")) \(policies.updated)")
                        .accessibilityIdentifier("policy-updated")
                    Text(policies.notice)
                    ForEach(page.sections.indices, id: \.self) { index in
                        Text(page.sections[index].title).font(.headline)
                        Text(page.sections[index].body).accessibilityIdentifier("policy-section-\(index)")
                    }
                    ForEach(page.sources.indices, id: \.self) { index in
                        Button {
                            guard let url = URL(string: page.sources[index].url) else { linkFailed = true; return }
                            openURL(url) { accepted in linkFailed = !accepted }
                        } label: {
                            VStack(alignment: .leading) {
                                Text(page.sources[index].label).font(.headline)
                                Text(page.sources[index].url)
                            }
                        }.accessibilityIdentifier("policy-source-\(index)")
                    }
                }.padding()
            }.accessibilityIdentifier("policy-body")
                .navigationTitle(page.title)
                .safeAreaInset(edge: .bottom) {
                    if linkFailed { Text(content.text("policyLinkFailed")).padding().accessibilityIdentifier("policy-link-failed") }
                }
                .toolbar { Button(content.text("close"), action: close).accessibilityIdentifier("policy-close") }
        }
    }
}

struct RegistrationView: View {
    @ObservedObject var model: AccountRegistration
    let content: MobileContent
    let policies: MobilePolicies
    @State private var name = ""
    @State private var email = ""
    @State private var password = ""
    @State private var code = ""
    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text(content.text("createAccount")).font(.title2.bold())
            if model.state.failed { Text(content.text("registrationFailed")) }
            switch model.state.step {
            case .details:
                TextField(content.text("fullName"), text: $name).textContentType(.name).disabled(model.state.busy)
                TextField(content.text("email"), text: $email).keyboardType(.emailAddress).textContentType(.username)
                    .textInputAutocapitalization(.never).autocorrectionDisabled().disabled(model.state.busy)
                SecureField(content.text("password"), text: $password).textContentType(.newPassword).disabled(model.state.busy)
                Text(policies.notice)
                PolicyLinks(policies: policies, content: content)
                Toggle(policies.acceptance, isOn: Binding(get: { model.state.accepted }, set: model.accept)).disabled(model.state.busy)
                Button(content.text("register")) {
                    let secret = password; password = ""; model.register(email: email, password: secret, name: name)
                }.disabled(model.state.busy || !model.state.accepted || email.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                           || name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || password.isEmpty)
            case .confirm:
                Text(content.text("registrationInstructions"))
                TextField(content.text("code"), text: $code).textContentType(.oneTimeCode).keyboardType(.numberPad).disabled(model.state.busy)
                Button(content.text("verify")) { let value = code; code = ""; model.confirm(value) }
                    .disabled(model.state.busy || code.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                Button(content.text("resendCode"), action: model.resend).disabled(model.state.busy)
                if model.state.codeRequested { Text(content.text("registrationCodeRequested")) }
            case .complete: Text(content.text("registrationComplete"))
            case .closed: EmptyView()
            }
            if model.state.busy { ProgressView(content.text("loading")) }
            Button(content.text("backToSignIn")) { name = ""; email = ""; password = ""; code = ""; model.close() }
        }.textFieldStyle(.roundedBorder).buttonStyle(.borderedProminent)
    }
}

File: native/android/app/src/main/java/com/mediconnect/nativeapp/RegistrationViews.kt
package com.mediconnect.nativeapp

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.Checkbox
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp

@Composable
fun PolicyLinks(policies: MobilePolicies, content: MobileContent) {
    var selected by remember { mutableStateOf<String?>(null) }
    Column {
        policies.pages.forEach { (key, page) ->
            TextButton(onClick = { selected = key }) { Text(page.title) }
        }
    }
    selected?.let { key ->
        val page = policies.pages.getValue(key)
        AlertDialog(onDismissRequest = { selected = null }, title = { Text(page.title) }, text = {
            LazyColumn(Modifier.heightIn(max = 480.dp), verticalArrangement = Arrangement.spacedBy(16.dp)) {
                item { Text(policies.notice) }
                page.sections.forEach { section -> item {
                    Column { Text(section.title, style = MaterialTheme.typography.titleSmall); Text(section.body) }
                } }
                page.sources.forEach { source -> item { Column { Text(source.label); Text(source.url) } } }
            }
        }, confirmButton = { TextButton(onClick = { selected = null }) { Text(content.text("close")) } })
    }
}

@Composable
fun RegistrationForm(state: RegistrationState, content: MobileContent, policies: MobilePolicies,
                     accept: (Boolean) -> Unit, register: (String, String, String) -> Unit,
                     confirm: (String) -> Unit, resend: () -> Unit, close: () -> Unit) {
    var name by remember(state.step) { mutableStateOf("") }
    var email by remember(state.step) { mutableStateOf("") }
    var password by remember(state.step) { mutableStateOf("") }
    var code by remember(state.step) { mutableStateOf("") }
    Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
        Text(content.text("createAccount"), style = MaterialTheme.typography.headlineSmall)
        if (state.failed) Text(content.text("registrationFailed"), color = MaterialTheme.colorScheme.error)
        when (state.step) {
            RegistrationStep.DETAILS -> {
                OutlinedTextField(name, { name = it }, Modifier.fillMaxWidth(), enabled = !state.busy,
                    label = { Text(content.text("fullName")) }, singleLine = true)
                OutlinedTextField(email, { email = it }, Modifier.fillMaxWidth(), enabled = !state.busy,
                    label = { Text(content.text("email")) }, singleLine = true, keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Email))
                OutlinedTextField(password, { password = it }, Modifier.fillMaxWidth(), enabled = !state.busy,
                    label = { Text(content.text("password")) }, singleLine = true, visualTransformation = PasswordVisualTransformation(),
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password))
                Text(policies.notice)
                PolicyLinks(policies, content)
                Row {
                    Checkbox(state.accepted, accept, modifier = Modifier.semantics { contentDescription = policies.acceptance }, enabled = !state.busy)
                    Text(policies.acceptance)
                }
                Button(onClick = {
                    val secret = password; password = ""; register(email, secret, name)
                }, enabled = !state.busy && state.accepted && name.isNotBlank() && email.isNotBlank() && password.isNotEmpty()) {
                    Text(content.text("register"))
                }
            }
            RegistrationStep.CONFIRM -> {
                Text(content.text("registrationInstructions"))
                OutlinedTextField(code, { code = it }, Modifier.fillMaxWidth(), enabled = !state.busy,
                    label = { Text(content.text("code")) }, singleLine = true, keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.NumberPassword))
                Button(onClick = { val value = code; code = ""; confirm(value) }, enabled = !state.busy && code.isNotBlank()) {
                    Text(content.text("verify"))
                }
                TextButton(onClick = resend, enabled = !state.busy) { Text(content.text("resendCode")) }
                if (state.codeRequested) Text(content.text("registrationCodeRequested"))
            }
            RegistrationStep.COMPLETE -> Text(content.text("registrationComplete"))
            RegistrationStep.CLOSED -> Unit
        }
        if (state.busy) CircularProgressIndicator()
        TextButton(onClick = { name = ""; email = ""; password = ""; code = ""; close() }) { Text(content.text("backToSignIn")) }
    }
}
